wake/testing/fuzzing/fuzz_shrink.py

```diff
@@ -0,0 +1,759 @@
+from __future__ import annotations
+
+
+from collections import defaultdict
+from typing import Callable, DefaultDict, List, Optional, Any, Tuple
+
+from typing_extensions import get_type_hints
+
+from wake.development.globals import random
+
+from ..core import get_connected_chains
+from .generators import generate
+
+from dataclasses import dataclass
+
+from .fuzz_test import FuzzTest
+
+from wake.development.globals import random, set_sequence_initial_internal_state, get_fuzz_mode, get_sequence_initial_internal_state, set_error_flow_num, get_error_flow_num, get_config, get_shrank_path
+
+from wake.development.core import Chain
+import pickle
+
+from pathlib import Path
+from datetime import datetime, timedelta
+
+import traceback
+from wake.utils.file_utils import is_relative_to
+from wake.development.transactions import Error
+import copy
+import os
+import sys
+from wake.cli.console import console
+from contextlib import contextmanager, redirect_stdout, redirect_stderr
+
+
+IGNORE_FLOW_INDEX = True # True if you accept it could reproduce same error earlier.
+
+def __get_methods(target, attr: str) -> List[Callable]:
+    ret = []
+    for x in dir(target):
+        if hasattr(target.__class__, x):
+            m = getattr(target.__class__, x)
+            if hasattr(m, attr) and getattr(m, attr):
+                ret.append(m)
+    return ret
+
+
+def clear_previous_lines(num_lines):
+    for _ in range(num_lines):
+        sys.stdout.write("\033[F")  # Move cursor up one line
+        sys.stdout.write("\033[K")  # Clear the line
+
+def compare_exceptions(e1, e2):
+    if type(e1) != type(e2):
+        return False
+
+    # from wake.development.transactions import Error
+    if type(e1) == Error and type(e2) == Error:
+        # If it was the Error(TransactionRevertedError), compare message content.
+        if e1.message != e2.message:
+            return False
+
+    tb1 = traceback.extract_tb(e1.__traceback__)
+    tb2 = traceback.extract_tb(e2.__traceback__)
+
+    frame1 = None
+    for frame1 in tb1:
+        if is_relative_to(
+            Path(frame1.filename), Path.cwd()
+        ) and not is_relative_to(
+            Path(frame1.filename), Path().cwd() / "pytypes"
+        ):
+            break
+    frame2 = None
+    for frame2 in tb2:
+        if is_relative_to(
+            Path(frame2.filename), Path.cwd()
+        ) and not is_relative_to(
+            Path(frame2.filename), Path().cwd() / "pytypes"
+        ):
+            break
```

Comment:
Yes, exactly. 
If it was a transaction error, rely on the Error type also if transaction. Error then relies on message. And does not care about another error argument.

If it was an error in Python, rely on the file and line number in Python. so we do not care about the actual value.

There was a lot of consideration.
Since it could depend on the definition of the "same error".

The purpose of shrinking is to create a minimum flow sequence to reproduce the same error.

The same error could be different.
1. the same error is emitted.
2. the same assertion in Python fails. (fails at the same line in the Python test) 

or 

3. the same error with the same argument is emitted. (and also in exactly the same flow)
4. the same assertion in Python fails with the same value. (and also in exactly the same flow)


I decided to implement 1. and 2. since these conditions could significantly shorten the test.
For example, If it was 3. 4. and Error was emitted with `TransferError(nft_id=10)`, then at least the test required to emit 10 NFT and make an error. This would be redundant.

However, one possible issue is that when checking the balance for each account, the shrinken result shows an unbalance for different accounts.
```python
   @invariant(period=30)
    def invariant_erc20_balances(self):
        for contract in self.erc20_balances:
            for acc in self.erc20_balances[contract]: 
                assert contract.balanceOf(acc) == self.erc20_balances[contract][acc] # <- error in same file and same line.
```

Reply:
Got it. Then we should at least implement the same logic for Panic error as it behaves the same as Error.

Still I think we should implement strict shrinking feature where we compare the errors exactly with `==`.

Reply:
Yes, that's true.

I already have an exact match for the flow number. I can extend this also Error matching.

https://github.com/Ackee-Blockchain/wake/blob/6256be27f3ea5e29a2329563a3a135e61261b14f/wake/testing/fuzzing/fuzz_shrink.py#L36